primitives/header-chain/src/lib.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Defines traits which represent a common interface for Substrate pallets which want to
+//! incorporate bridge functinality.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::dispatch::DispatchResult;
+use sp_std::prelude::*;
+
+/// A base trait for pallets which want to keep track of a full set of headers from a bridged chain.
+pub trait MinimalHeaderChain<AccountId> {
+	/// The header used by the chain.
+	type Header;
+	/// Any extra data which helps describe a transaction.
+	type Extra;
+	/// The type of block number used by the chain.
+	type BlockNumber;
+	/// The type of block hash used by the chain.
+	type BlockHash;
+
+	/// Imports a header submitted using an _unsigned_ transaction to the pallet.
+	fn import_header_unsigned(header: Self::Header, extra_data: Option<Self::Extra>) -> DispatchResult;
+
+	/// Imports a header submitted using a _signed_ transaction to the pallet.
+	fn import_header_signed(
+		submitter: AccountId,
+		header: Self::Header,
+		extra_data: Option<Self::Extra>,
+	) -> DispatchResult;
+
+	/// Get the best block the pallet knows of.
+	///
+	/// May return multiple blocks if we have unfinalized blocks on different forks
+	/// which are at the same height.
+	fn best_headers() -> Vec<Self::Header>;
```

Comment:
What would be the use case for this API? How would bridge application use this information?

Reply:
The use case here would be to see if the block you're about to submit is newer than the best block the bridge knows of. If is it not newer then it would not be a good idea to submit it since it could get flagged as "useless" during the signed header import. We don't do anything with this information at the moment, but further down the line this could be used as a sort of reputation mechanism for peers

Reply:
Actually, at the moment that can be accomplished with `best_block()`, so maybe `best_headers()` is not required